README.md

```diff
@@ -1 +1,189 @@
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
+
+## 2.2 Raw (Noncanonicalized) Data
+
+Canonicalization is not required if data is encoded as raw bytes (multicodec `0x00`). The exact bytes are already present, and MUST not be changed.
+
+# 3 Varsig Format
+
+A varsig is a bytestring that includes the following information:
+
+| Segment Name              | Type     | Description                                     | Required |
+|---------------------------|----------|-------------------------------------------------|----------|
+| Varsig Prefix             | `0xd000` | The multicodec varsig prefix                    | Yes      |
+| Public Key Prefix         | `Varint` | The multicodec prefix for the public key type   | Yes      |
+| Hash Prefix               | `Varint` | The multicodec prefix for the hash algorithm    | Yes      |
+| Hash Length               | `Varint` | The hash length                                 | Yes      |
+| Content Multicodec Prefix | `Varint` | The IPLD encoding uses to canonicalize the data | Yes      |
+| Raw Signature             | `Varint` | The raw signature                               | Yes      |
+  
+## 3.1 Segments
+
+### 3.1.1 Varsig Prefix
+
+The varsig prefix MUST be `0xd000`.
+
+### 3.1.2 Public Key Prefix
+
+The [multicodec](https://github.com/multiformats/multicodec) prefix for the public key type associated with the signature.
+
+For example: `0xe7` for secp256k1, or `0x1205` for RSA.
+
+### 3.1.3 Hash Prefix
+
+The [multicodec](https://github.com/multiformats/multicodec) hash prefix used by the signature. This is the first segment of a [multihash](https://github.com/multiformats/multihash).
+
+For example: `0x16` for SHA3-256, or `0x1e` for BLAKE3
+
+### 3.1.4 Hash Length
+
+The hash length used by the signature. This is the second segment of a [multihash](https://github.com/multiformats/multihash).
+
+### 3.1.5 Content Multicodec Prefix
+
+The [multicodec](https://github.com/multiformats/multicodec) prefix of the encoding scheme used.
+
+For example: `0x00` for raw bytes (no special encoding), `0x0129` for DAG-JSON, and `0x70` for DAG-PB.
+
+### 3.1.6 Raw Signature
+
+The raw signature bytes
+
+## 3.2 Byte Segments
+
+```xml
+<varint 0xd000><varint multicodec_key_prefix><varint multicodec_hash_prefix><varint multicodec_hash_length><varint multicodec_prefix><varint raw_hash><varint raw_signature>
+```
+
+## 3.3 IPLD Schema
+
+``` ipldsch
+type TaggedVarsig union {
+  | Varsig "d000"
+} representation bytesprefix
+
+type Varsig struct {
+  keyPrefix Multicodec.PublicKey
+  hashPrefix Multicodec.Multihash
+  digestLength Integer
+  contentEnc Multicodec
+  rawSig Bytes
```

Comment:
How do we know the length of the signature? Is it implied by the `keyPrefix`?

Reply:
> How do we know the length of the signature? Is it implied by the keyPrefix?

I think that is what `digestLength` is

Reply:
Oh, I thought that was the length for the hash function used, but this makes more sense!

Reply:
We could also rely on the size of the raw signature itself, though would require computation at runtime to count the signature bytes.

Reply:
> Oh, I thought that was the length for the hash function used, but this makes more sense!

Nevermind, I totally misunderstood the question. Yes, the signature should be a varint.

The digest length is [lifted directly from multihash](https://github.com/multiformats/multihash#different-hash-functions):

![](https://raw.githubusercontent.com/multiformats/multihash/master/img/multihash.006.jpg)

This is because the hash can be truncated:

![](https://raw.githubusercontent.com/multiformats/multihash/master/img/multihash.004.jpg)

to your _original_ question: yes, the raw signature should be a varint 👍 Will fix!

Reply:
After looking at the multicodec table in more depth, I changed it to just use the multihashes in multicodec.

Reply:
It seems like both approaches are workable?
1. Either a varint describing the length of the signature
2. Or make the signature itself a varint

To me option (1) seems cleaner? 
It's also more in line with multihash!

Reply:
@QuinnWilton I think may be what you're saying in your remaining unresolved comment?

Reply:
I wonder if a varint is just the easiest thing here. With the separate field, there's one more thing that can go wrong for no good reason (claims that the signature is 2048-bits, but is actually 1024). There's less conversion with the bytes, but we end up with a varint either way (the signature itself or the signature length field).

But that's just me 🤷‍♀️